pkg/volume/azure_dd/provisioner.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure_dd
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type azureDiskProvisioner struct {
+	plugin  *azureDataDiskPlugin
+	options volume.VolumeOptions
+}
+
+type azureDiskDeleter struct {
+	*dataDisk
+	spec   *volume.Spec
+	plugin *azureDataDiskPlugin
+}
+
+var _ volume.Provisioner = &azureDiskProvisioner{}
+var _ volume.Deleter = &azureDiskDeleter{}
+
+func (d *azureDiskDeleter) GetPath() string {
+	return getPath(d.podUID, d.dataDisk.diskName, d.plugin.host)
+}
+
+func (d *azureDiskDeleter) Delete() error {
+	volumeSource, err := getVolumeSource(d.spec)
+	if err != nil {
+		return err
+	}
+
+	wasStandAlone := (*volumeSource.Kind != v1.AzureSharedBlobDisk)
+	managed := (*volumeSource.Kind == v1.AzureManagedDisk)
+
+	if managed {
+		return d.plugin.diskController.DeleteManagedDisk(volumeSource.DataDiskURI)
+	}
+
+	return d.plugin.diskController.DeleteBlobDisk(volumeSource.DataDiskURI, wasStandAlone)
+}
+
+func (p *azureDiskProvisioner) Provision() (*v1.PersistentVolume, error) {
+	supportedModes := p.plugin.GetAccessModes()
+
+	// perform static validation first
+	if p.options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("azureDisk - claim.Spec.Selector is not supported for dynamic provisioning on Azure disk")
+	}
+
+	if len(p.options.PVC.Spec.AccessModes) > 1 {
```

Comment:
Yes. It is not elegant. Verbosity meant to help the user understand what went wrong with every case instead of cryptic error messages.